hubblestack/extmods/modules/audit.py

```diff
@@ -1,450 +1,726 @@
-    files for this host and execute and return the results of those audit files.
-
-    .. code-block:: yaml
-
-        audit:
-          '*':
-            - cis.linux
-            - adobebaseline
-          'G@os:CoreOS':
-            - coreos_additional_checks
-
-    :param topfile:
-        Topfile to process for targeted audit files to execute.
-    :param tags:
-        See audit()
-    :param labels:
-        See audit()
+    Function to execute the module and return the result
+    :param audit_id:
+    :param audit_impl:
+    :param audit_data:
     :param verbose:
-        See audit()
-    :param show_success:
-        See audit()
+    :param nova_profile:
     :return:
-        Returns dictionary with Success and Failure keys and the results of the
-        checks
     """
-    audit_files = _get_top_data(topfile)
+    audit_result = {
+        "check_id": audit_id,
+        "description": audit_data['description'],
+        "nova_profile": nova_profile,
+        "sub_check": audit_data.get('sub_check', False),
+        "tag": audit_data['tag'],
+        "module": audit_impl['module'],
+        "run_config": {
+            "filter": audit_impl['filter'],
+        }
+    }
+
+    failure_reason = audit_data.get('failure_reason', '')
+    invert_result = audit_data.get('invert_result', False)
+    return_no_exec = audit_impl.get('return_no_exec', False)
+    type = audit_impl.get('type', 'and').lower().strip()
+
+    # check for type in check implementation. If not present default is 'and'
+    audit_result['run_config']['type'] = type
+    audit_result['invert_result'] = invert_result
+
+    # check if return_no_exec is true
+    if return_no_exec:
+        audit_result['run_config']['return_no_exec'] = True
+        check_result = CHECK_STATUS['Success']
+        if invert_result:
+            check_result = CHECK_STATUS['Failure']
+            audit_result['failure_reason'] = failure_reason
+        audit_result['check_result'] = check_result
+        return audit_result
+
+    # Check presence of implementation checks
+    if 'checks' not in audit_impl:
+        raise AuditCheckValidationError('No checks are present')
+
+    # Execute module validation of params
+    validate_param_method = audit_impl['module'] + '.validate_params'
+    for audit_check in audit_impl['checks']:
+        __nova__[validate_param_method](audit_id, audit_check)
+
+    # validate succeeded, lets execute it and prepare result dictionary
+    audit_result['run_config']['checks'] = []
+    execute_method = audit_impl['module'] + '.execute'
+    filtered_log_method = audit_impl['module'] + '.get_filtered_params_to_log'
+    # calculate the check result based on type parameter.
+    # If type is 'and', all subchecks should pass for success.
+    # If type is 'or', any passed subcheck will result in success.
+    overall_result = type == 'and'
+    failure_reasons = []
+    for audit_check in audit_impl['checks']:
+        if result_list is not None:
```

Comment:
Let's add some comments here to indicate why we're doing this. (For bexpr checks.)